### ISSUE TITLE
Error during the documentation's "compiling substrate" section i.e. the `make build` command.

### ISSUE THREAD

**Issue by tpmccallum** [substrate-developer-hub/substrate-developer-hub.github.io#908]
Hi,
I am receiving an error when executing the `make build` command at the following Substrate Developer Hub documentation < https://substrate.dev/docs/en/tutorials/create-your-first-substrate-chain/setup#compiling-substrate >
```
Compiling static_assertions v1.1.0
error[E0210]: type parameter `T` must be used as the type parameter for some local type (e.g., `MyStruct<T>`)
   --> /Users/tpmccallum/.cargo/registry/src/github.com-1ecc6299db9ec823/subtle-2.4.0/src/lib.rs:513:1
    |
513 | impl<T> From<CtOption<T>> for Option<T> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type parameter `T` must be used as the type parameter for some local type
    |
    = note: only traits defined in the current crate can be implemented for a type parameter

error: aborting due to previous error

For more information about this error, try `rustc --explain E0210`.
error: could not compile `subtle`.
warning: build failed, waiting for other jobs to finish...
error: build failed
make: *** [build] Error 101
```
I am on macOS Big Sur, and all other steps have succeeded to this point.

Thank you
Tim

**Comment by NukeManDan** [comment]
Tim - try to start with [v3 template](https://github.com/substrate-developer-hub/substrate-node-template/tree/v3.0.0), the documentation is being updated for this presently, the makefile will go away, so instead, try running `cargo build` with a completely fresh v3 git clone 🙏🏼

**Comment by NukeManDan** [comment]
@tomsik68 #911 closes this, as the v3 create your first chain tutorial is now confirmed working 😁

**Comment by tpmccallum** [comment]
Hi @NukeManDan,
Sorry to bother you, I am still getting the same error :(

Here are the details
```
# Install Homebrew if necessary https://brew.sh/
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"

# Make sure Homebrew is up-to-date, install openssl and cmake
brew update
brew install openssl cmake

# Install Rust
curl https://sh.rustup.rs -sSf | sh

# Configure Rust
source ~/.cargo/env

# Configure toolchain
rustup default stable
rustup update nightly
rustup update stable
rustup target add wasm32-unknown-unknown --toolchain nightly
```

# Get substrate node template
```
git clone -b v3.0.0 https://github.com/substrate-developer-hub/substrate-node-template.git
```

# Use Rust's cargo command to build and launch
```
cd substrate-node-template
cargo run --release -- --dev --tmp
```

At present, the above results in `error: build failed`. See below 

## Error

```
error[E0210]: type parameter `T` must be used as the type parameter for some local type (e.g., `MyStruct<T>`)
   --> /Users/tpmccallum/.cargo/registry/src/github.com-1ecc6299db9ec823/subtle-2.4.0/src/lib.rs:513:1
    |
513 | impl<T> From<CtOption<T>> for Option<T> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type parameter `T` must be used as the type parameter for some local type
    |
    = note: only traits defined in the current crate can be implemented for a type parameter

error: aborting due to previous error

For more information about this error, try `rustc --explain E0210`.
   Compiling block-padding v0.2.1
error: could not compile `subtle`.
warning: build failed, waiting for other jobs to finish...
error: build failed
```

I tried using a fresh git clone of v3 [based on this help/feedback](https://github.com/substrate-developer-hub/substrate-developer-hub.github.io/issues/908#issuecomment-799753885), however I am still getting the above error.

I can confirm that we are using v3.0.0 by issuing the following command.
```
tag=$(git describe --tags `git rev-list --tags --max-count=1`)
echo $tag
```
Result
```
v3.0.0
```

**Comment by tpmccallum** [comment]
![Uploading Screen Shot 2021-03-18 at 3.12.51 pm.png…]()

**Comment by NukeManDan** [comment]
Bummer! I will try to confirm this behavior but don't have a Mac, and all reports of Linux builds seem to work.

Can you share your machine spec?

**Comment by tpmccallum** [comment]
![Screen Shot 2021-03-18 at 4 26 58 pm](https://user-images.githubusercontent.com/9831342/111582614-d2a2d480-8806-11eb-83bc-b75592f28849.png)

**Comment by tpmccallum** [comment]
So just to reconfirm, I tried the installation from this URL also < https://substrate.dev/docs/en/tutorials/create-your-first-substrate-chain/setup >

Specifically, these steps ...
```
git clone -b v3.0.0 --depth 1 https://github.com/substrate-developer-hub/substrate-node-template
cd substrate-node-template
cargo build --release
```
Achieved same outcome, as shown below.
```
   Compiling libc v0.2.82
   Compiling proc-macro2 v1.0.24
   Compiling unicode-xid v0.2.1
   Compiling syn v1.0.58
   Compiling cfg-if v1.0.0
   Compiling cfg-if v0.1.10
   Compiling autocfg v1.0.1
   Compiling log v0.4.13
   Compiling serde_derive v1.0.119
   Compiling serde v1.0.119
   Compiling byteorder v1.4.2
   Compiling lazy_static v1.4.0
   Compiling scopeguard v1.1.0
   Compiling smallvec v1.6.1
   Compiling version_check v0.9.2
   Compiling memchr v2.3.4
   Compiling typenum v1.12.0
   Compiling ppv-lite86 v0.2.10
   Compiling ahash v0.4.7
   Compiling futures v0.1.30
   Compiling either v1.6.1
   Compiling slab v0.4.2
   Compiling getrandom v0.2.1
   Compiling futures-core v0.3.11
   Compiling pin-project-lite v0.2.4
   Compiling anyhow v1.0.38
   Compiling getrandom v0.1.16
   Compiling futures-io v0.3.11
   Compiling proc-macro-hack v0.5.19
   Compiling futures-sink v0.3.11
   Compiling arrayvec v0.5.2
   Compiling proc-macro-nested v0.1.7
   Compiling subtle v2.4.0
   Compiling pin-utils v0.1.0
   Compiling opaque-debug v0.3.0
error[E0210]: type parameter `T` must be used as the type parameter for some local type (e.g., `MyStruct<T>`)
   --> /Users/tpmccallum/.cargo/registry/src/github.com-1ecc6299db9ec823/subtle-2.4.0/src/lib.rs:513:1
    |
513 | impl<T> From<CtOption<T>> for Option<T> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type parameter `T` must be used as the type parameter for some local type
    |
    = note: only traits defined in the current crate can be implemented for a type parameter

error: aborting due to previous error

For more information about this error, try `rustc --explain E0210`.
error: could not compile `subtle`.
warning: build failed, waiting for other jobs to finish...
error: build failed

```

**Comment by tpmccallum** [comment]
Oh snap @NukeManDan it is working now.

I installed the following

```
brew install python@3.9 protobuf
```

And then was able to run the following successfully
```
cargo build --release
```

The solution was thanks to [this article](https://vikiival.medium.com/run-substrate-on-apple-m1-a2699743fae8) which is actually linked to from the official [substrate development documentation](https://substrate.dev/docs/en/tutorials/create-your-first-substrate-chain/setup#substrate-development).

Perhaps we could add the `brew install python@3.9 protobuf` to the macOS doc section at [v3's macOS section](https://github.com/substrate-developer-hub/substrate-node-template/blob/v3.0.0/doc/rust-setup.md#macos)

Thanks so much, sorry to use up your valuable time.
Tim

**Comment by NukeManDan** [comment]
Glad this worked! I see your rust toolchain is very old though, are you sure this could not have lead to the issue? rustc 1.5 and up is what we are suggesting

`rustup update`

**Comment by NukeManDan** [comment]
I will also inquire to ask other Mac users to confirm the brew deps required 🙏🏼

**Comment by tpmccallum** [comment]
Hi @NukeManDan 
A thousand apologies ... I fired up a new MacBook which has rustc 1.50.0 and it worked. It seems that you are correct in saying that the toolchain etc need to be fresh in order for the compilation to work.

I could have sworn that I performed the following commands on that other system where the error occured i.e.

```
rustup update nightly
rustup update stable
```

This makes me wonder if `rustup update` (as a stand-alone command) performs a different action to either `rustup update nightly` and/or `rustup update stable`. I have since returned to the other system and run `rustup update` and now the issues is fixed on that other system.

I would need to get a brand new fresh system to test this out. For now just letting you know that everything is working as expected.
